subsys/bluetooth/controller/ll_sw/ull_llcp.c

```diff
@@ -0,0 +1,470 @@
+		LL_ASSERT(0);
+	}
+
+	lp_comm_execute_fsm(conn, ctx, LP_COMMON_EVT_PREPARE);
+}
+
+static void lr_act_complete(struct ull_cp_conn *conn)
+{
+	/* Dequeue pending connuest that just completed */
+	(void) lr_dequeue(conn);
+}
+
+static void lr_act_connect(struct ull_cp_conn *conn)
+{
+	/* TODO */
+}
+
+static void lr_act_disconnect(struct ull_cp_conn *conn)
+{
+	lr_dequeue(conn);
+}
+
+static void lr_st_disconnect(struct ull_cp_conn *conn, u8_t evt)
+{
+	switch (evt) {
+	case LR_EVT_CONNECT:
+		lr_act_connect(conn);
+		conn->local.state = LR_STATE_IDLE;
+		break;
+	default:
+		/* Ignore other evts */
+		break;
+	}
+}
+
+static void lr_st_idle(struct ull_cp_conn *conn, u8_t evt)
+{
+	switch (evt) {
+	case LR_EVT_PREPARE:
+		if (lr_peek(conn)) {
+			lr_act_prepare(conn);
+			conn->local.state = LR_STATE_ACTIVE;
+		}
+		break;
+	case LR_EVT_DISCONNECT:
+		lr_act_disconnect(conn);
+		conn->local.state = LR_STATE_DISCONNECT;
+		break;
+	default:
+		/* Ignore other evts */
+		break;
+	}
+}
+
+static void lr_st_active(struct ull_cp_conn *conn, u8_t evt)
+{
+	switch (evt) {
+	case LR_EVT_COMPLETE:
+		lr_act_complete(conn);
+		conn->local.state = LR_STATE_IDLE;
+		break;
+	case LR_EVT_DISCONNECT:
+		lr_act_disconnect(conn);
+		conn->local.state = LR_STATE_DISCONNECT;
+		break;
+	default:
+		/* Ignore other evts */
+		break;
+	}
+}
+
+static void lr_execute_fsm(struct ull_cp_conn *conn, u8_t evt)
+{
+	switch (conn->local.state) {
+	case LR_STATE_DISCONNECT:
+		lr_st_disconnect(conn, evt);
+		break;
+	case LR_STATE_IDLE:
+		lr_st_idle(conn, evt);
+		break;
+	case LR_STATE_ACTIVE:
+		lr_st_active(conn, evt);
+		break;
+	default:
+		/* Unknown state */
+		LL_ASSERT(0);
+	}
+}
+
+static void lr_prepare(struct ull_cp_conn *conn)
+{
+	lr_execute_fsm(conn, LR_EVT_PREPARE);
+}
+
+static void lr_complete(struct ull_cp_conn *conn)
+{
+	lr_execute_fsm(conn, LR_EVT_COMPLETE);
+}
+
+static void lr_connect(struct ull_cp_conn *conn)
+{
+	lr_execute_fsm(conn, LR_EVT_CONNECT);
+}
+
+static void lr_disconnect(struct ull_cp_conn *conn)
+{
+	lr_execute_fsm(conn, LR_EVT_DISCONNECT);
+}
+
+/*
+ * LLCP Procedure Creation
+ */
+
+static struct proc_ctx *create_procedure(u8_t proc)
+{
+	struct proc_ctx *ctx;
+
+	ctx = proc_ctx_acquire();
+	if (!ctx) {
+		return NULL;
+	}
+
+	ctx->proc = proc;
+	ctx->state = LP_COMMON_STATE_IDLE;
+	ctx->collision = 0U;
+	ctx->pause = 0U;
+
+	return ctx;
+}
+
+/*
+ * LLCP Procedure Helper
+ */
+
+static int version_exchange(struct ull_cp_conn *conn)
+{
+	struct proc_ctx *ctx;
+
+	ctx = create_procedure(PROC_VERSION_EXCHANGE);
+	if (!ctx) {
+		return BT_HCI_ERR_CMD_DISALLOWED;
+	}
+
+	lr_enqueue(conn, ctx);
+
+	return BT_HCI_ERR_SUCCESS;
+}
+
+/*
+ * LLCP Public API
+ */
+
+void ull_cp_init(void)
+{
+	/**/
+	mem_init(mem_ctx.pool, sz_mem_ctx, ncount_mem_ctx, &mem_ctx.free);
+	mem_init(mem_tx.pool, sz_mem_tx, ncount_mem_tx, &mem_tx.free);
+}
+
+void ull_cp_conn_init(struct ull_cp_conn *conn)
+{
+	/**/
+	conn->local.state = LR_STATE_DISCONNECT;
```

Comment:
LR_STATE_DISCONNECT? Shouldn't it be something LR_STATE_IDLE?

Reply:
`LR_STATE_DISCONNECT` is a state yo make sure that you cannot perform any more interaction before you bring it back into `LR_STATE_IDLE`.

Reply:
ok